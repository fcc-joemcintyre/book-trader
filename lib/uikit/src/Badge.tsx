import styled from 'styled-components';

export type TBadge = {
  c?: string,
  tc?: string,
  bg?: string,
  tbg?: string,
};

export const Badge = styled.div<TBadge>`
  display: inline-block;
  padding: 1px 8px;
  ${({ c, tc, bg, tbg, theme }) => `
    color: ${(tc && theme.colors[tc]) ? theme.colors[tc] : c || '#ffffff'};
    background-color: ${(tbg && theme.colors[tbg]) ? theme.colors[tbg] : bg || '#0000ff'};
  `}
  border-radius: 999px;
`;
