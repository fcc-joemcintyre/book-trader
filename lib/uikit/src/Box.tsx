import styled from 'styled-components';

const fontsize: Record<string, { index: number, default: string}> = {
  xxsmall: { index: 0, default: '11px' },
  xsmall: { index: 1, default: '12px' },
  small: { index: 2, default: '14px' },
  normal: { index: 3, default: '16px' },
  large: { index: 4, default: '18px' },
  xlarge: { index: 5, default: '22px' },
  xxlarge: { index: 6, default: '26px' },
};

export type TBox = {
  font?: string,
  fs?: string,
  m?: string,
  mt?: string,
  mb?: string,
  ml?: string,
  mr?: string,
  p?: string,
  pt?: string,
  pb?: string,
  pl?: string,
  pr?: string,
  h?: string,
  minh?: string,
  maxh?: string,
  w?: string,
  minw?: string,
  maxw?: string,
  c?: string,
  tc?: string,
  bg?: string,
  tbg?: string,
  b?: string,
  br?: string,
  inline?: boolean,
  center?: boolean,
  align?: string,
  cursor?: string,
  overflow?: string,
}

export const Box = styled.div<TBox>`
  ${({ theme, font }) => font && ((theme.fonts && theme.fonts[font]) ? theme.fonts[font] :
    theme.fonts && theme.fonts.global ? theme.fonts.global : `
    font-family: sans-serif;
    font-weight: normal;
  `)};
  ${({ theme, fs }) => (fs && fontsize[fs] ? `
    font-size: ${(theme.fontSize && theme.fontSize[fontsize[fs].index]) || fontsize[fs].default};
  ` : fs && `
    font-size: ${fs};
  `)};
  ${({ m, mt, mb, ml, mr, p, pt, pb, pl, pr, h, maxh, minh, w, maxw, minw,
    c, tc, bg, tbg, b, br, inline, center, align, cursor, overflow, theme }) => `
    ${m ? `margin: ${m};` : ''}
    ${mt ? `margin-top: ${mt};` : ''}
    ${mb ? `margin-bottom: ${mb};` : ''}
    ${ml ? `margin-left: ${ml};` : ''}
    ${mr ? `margin-right: ${mr};` : ''}
    ${p ? `padding: ${p};` : ''}
    ${pt ? `padding-top: ${pt};` : ''}
    ${pb ? `padding-bottom: ${pb};` : ''}
    ${pl ? `padding-left: ${pl};` : ''}
    ${pr ? `padding-right: ${pr};` : ''}
    ${h ? `height: ${h};` : ''}
    ${maxh ? `max-height: ${maxh};` : ''}
    ${minh ? `min-height: ${minh};` : ''}
    ${w ? `width: ${w};` : ''}
    ${maxw ? `max-width: ${maxw};` : ''}
    ${minw ? `min-width: ${minw};` : ''}
    ${c ? `color: ${c};` : ''}
    ${tc && theme.colors[tc] ? `color: ${theme.colors[tc]};` : ''}
    ${bg ? `background-color: ${bg};` : ''}
    ${tbg && theme.colors[tbg] ? `background-color: ${theme.colors[tbg]};` : ''}
    ${b ? `border: ${b};` : ''}
    ${br ? `border-radius: ${br};` : ''}
    ${inline ? 'display: inline-block;' : ''}
    ${center ? `
      margin-left: auto;
      margin-right: auto;
    ` : ''}
    ${align ? `text-align: ${align};` : ''}
    ${cursor ? `cursor: ${cursor};` : ''}
    ${overflow ? `overflow: ${overflow};` : ''}
  `}
`;
