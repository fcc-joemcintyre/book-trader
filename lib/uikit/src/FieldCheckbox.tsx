import styled from 'styled-components';
import { Field } from '@cygns/use-fields';
import { GridBoxElement } from './GridBoxElement';
import { FieldInfo } from './FieldInfo';
import { FieldError } from './FieldError';
import { FieldElementStyle } from './FieldElementStyle';
import { Text } from './Text';

export type Props = {
  field: Field,
  label?: string,
  span?: number,
  row?: boolean,
  info?: string,
  errors?: Record<string, string>,
  showInfo?: boolean,
  showErrors?: boolean,
  onChange: React.ChangeEventHandler,
  onValidate?: React.FocusEventHandler,
};

const StyledCheckbox = styled.div`
  ${FieldElementStyle}
  background-color: inherit;
  border: 1px solid transparent;
`;

export const FieldCheckbox = ({
  field,
  label = '',
  span = 12,
  row = false,
  info = undefined,
  errors = undefined,
  showInfo = true,
  showErrors = true,
  onChange,
  onValidate = () => { /* no op */ },
  ...rest
}: Props) => (
  <GridBoxElement span={span} row={row}>
    <StyledCheckbox>
      <input
        type='checkbox'
        {...rest}
        id={field.name}
        name={field.name}
        checked={field.value === true || field.value === 'true'}
        onChange={onChange}
        onBlur={onValidate}
      />
      <Text as='label' htmlFor={field.name}>{label}</Text>
    </StyledCheckbox>
    { showErrors && field.error && (
      <FieldError>
        { (field.error === 'required') ? 'Is required' :
            errors ? (errors[field.error] || 'Error') : 'Error'
        }
      </FieldError>
    )}
    { showInfo && (!field.error) && info && (
      <FieldInfo>
        { (info && (info.trim () !== '')) ? info : <span>&nbsp;</span>}
      </FieldInfo>
    )}
  </GridBoxElement>
);
