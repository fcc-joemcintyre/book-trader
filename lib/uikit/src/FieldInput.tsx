import React from 'react';
import styled from 'styled-components';
import { Field } from '@cygns/use-fields';
import { GridBoxElement } from './GridBoxElement.js';
import { FieldInfo } from './FieldInfo.js';
import { FieldError } from './FieldError.js';
import { FieldLabel } from './FieldLabel.js';
import { FieldElementStyle } from './FieldElementStyle.js';

type Props = {
  field: Field,
  label?: string,
  span?: number,
  row?: boolean,
  info?: string,
  errors?: Record<string, string>,
  showLabel?: boolean,
  showInfo?: boolean,
  showErrors?: boolean,
  onChange: React.ChangeEventHandler,
  onValidate?: React.FocusEventHandler,
};

const StyledInput = styled.input`
  ${FieldElementStyle}
  font-size: 16px;
  padding: 4px;
  outline: 0;
`;

export const FieldInput = ({
  field,
  label = '',
  span = 12,
  row = false,
  info = undefined,
  errors = undefined,
  showLabel = true,
  showInfo = true,
  showErrors = true,
  onChange,
  onValidate = () => { /* no op */ },
  ...rest
}: Props) => {
  const value: string = field.value as string;
  return (
    <GridBoxElement span={span} row={row}>
      {showLabel && <FieldLabel htmlFor={field.name} required={field.required}>{label}</FieldLabel>}
      <StyledInput
        type='text'
        {...rest}
        id={field.name}
        name={field.name}
        value={value}
        error={field.error ? true : false /* eslint-disable-line */}
        onChange={onChange}
        onBlur={onValidate}
      />
      { showErrors && field.error && (
        <FieldError>
          { (field.error === 'required') ? 'Is required' :
            errors ? (errors[field.error] || 'Error') : 'Error'
          }
        </FieldError>
      )}
      { showInfo && (!field.error) && (
        <FieldInfo>
          { (info && (info.trim () !== '')) ? info : <span>&nbsp;</span>}
        </FieldInfo>
      )}
    </GridBoxElement>
  );
};
