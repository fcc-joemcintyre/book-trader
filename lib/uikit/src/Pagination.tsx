import { Flex } from './Flex';
import { Text } from './Text';

type Props = {
  items: number,
  pageItems: number,
  visible: number,
  first: number,
  current: number,
  font: string,
  fs: string,
  c?: string,
  cselected?: string,
  tc?: string,
  tcselected?: string,
  gap?: string,
  onChange: (page: number, offset: number) => void,
};

export const Pagination = ({
  items,
  pageItems,
  visible,
  first,
  current,
  font,
  fs,
  c = '#000000',
  cselected = '#0000ff',
  tc,
  tcselected,
  gap = '6px',
  onChange,
  ...rest
}: Props) => {
  const pages = Math.floor (items / (pageItems > 0 ? pageItems : 1)) + (items % pageItems > 0 ? 1 : 0);
  const start = first < pages ? first : pages > 0 ? pages - 1 : 0;
  const end = Math.min (start + visible, pages) - 1;
  const page = current < start ? start : current > end ? end : current;
  const elements = [];
  for (let a = start; a <= end; a += 1) {
    elements.push (
      <Text
        key={a}
        font={font}
        fs={fs}
        c={a === page ? cselected : c}
        tc={a === page ? tcselected : tc}
        cursor='pointer'
        onClick={() => onChange (first, a)} // eslint-disable-line no-loop-func
      >
        {a === page ? <strong>{a + 1}</strong> : a + 1}
      </Text>
    );
  }
  return (
    <Flex gap={gap} {...rest}>
      { start > 0 ? (
        <Text
          font={font}
          fs={fs}
          c={c}
          tc={tc}
          cursor='pointer'
          onClick={() => {
            const c2 = Math.max (page - visible, 0);
            const f2 = Math.max (first - visible, 0);
            onChange (f2, c2);
          }}
        >
          &lt;
        </Text>
      ) : (
        <Text
          c='transparent'
          cursor='default'
        >
          &lt;
        </Text>
      )}
      {elements}
      { (end < (pages - 1)) && (
        <Text
          font={font}
          fs={fs}
          c={c}
          tc={tc}
          cursor='pointer'
          onClick={() => {
            const c2 = Math.min (page + visible, pages - 1);
            const f2 = Math.min (first + visible, pages - visible);
            onChange (f2, c2);
          }}
        >
          &gt;
        </Text>
      )}
    </Flex>
  );
};
