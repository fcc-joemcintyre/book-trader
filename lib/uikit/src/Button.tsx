import styled from 'styled-components';
import { Box, TBox } from './Box.js';

export type TButton = TBox & {
  as?: string,
  s?: string,
  v?: string,
  type?: string,
  bc?: string,
  hc?: string,
  hbg?: string,
  hbc?: string,
  dc?: string,
  dbg?: string,
  dbc?: string,
};

export const Button = styled (Box)<TButton>`
  border-radius: 4px;
  cursor: pointer;
  text-transform: uppercase;
  ${({ theme, v, type }) => {
    let base: string = v && theme.variant[v];
    if (!base) {
      base = type === 'submit' ? theme.variant.buttonSubmit : theme.variant.buttonDefault;
    }
    return base;
  }}
  ${({ theme, s, c, bg, bc, hc, hbg, hbc, dc, dbg, dbc }) => (`
    ${(s && theme.buttonSize[s]) || theme.buttonSize.normal}
    ${c && `color: ${c};`}
    ${bg && `background-color: ${bg};`}
    ${bc && `border-color: ${bc};`}

    &:hover {
      ${hc && `color: ${hc};`}
      ${hbg && `background-color: ${hbg};`}
      ${hbc && `border-color: ${hbc};`}
    }

    &:disabled {
      ${dc && `color: ${dc};`}
      ${dbg && `background-color: ${dbg};`}
      ${dbc && `border-color: ${dbc};`}
    }
  `)}

  ${({ s }) => (s === 'wide') && `
    width: 100%;
  `};
`;
