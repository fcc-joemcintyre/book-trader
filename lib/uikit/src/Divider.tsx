import styled from 'styled-components';
import { Box, TBox } from './Box';

export type TDivider = TBox & {
  extend: string,
};

export const Divider = styled (Box)<TDivider>`
  height: ${({ h }) => h || '1px'};
  background: ${({ c, tc, theme }) => ((tc && theme.colors[tc]) ? theme.colors[tc] : c || '#444444')};
  ${({ extend }) => extend && `
    margin-left: -${extend};
    margin-right: -${extend};
  `};
`;
