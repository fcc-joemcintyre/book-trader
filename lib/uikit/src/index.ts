import { Badge } from './Badge.js';
import { Box } from './Box.js';
import { Button } from './Button.js';
import { Divider } from './Divider.js';
import { FieldCheckbox } from './FieldCheckbox.js';
import { FieldElementStyle } from './FieldElementStyle.js';
import { FieldError } from './FieldError.js';
import { FieldInfo } from './FieldInfo.js';
import { FieldInput } from './FieldInput.js';
import { FieldLabel } from './FieldLabel.js';
import { FieldSelect } from './FieldSelect.js';
import { FieldText } from './FieldText.js';
import { FieldTextArea } from './FieldTextArea.js';
import { Flex } from './Flex.js';
import { FlexItem } from './FlexItem.js';
import { Font } from './Font.js';
import { FontSize } from './FontSize.js';
import { GridBox } from './GridBox.js';
import { GridBoxElement } from './GridBoxElement.js';
import { Image } from './Image.js';
import { LinkExternal } from './LinkExternal.js';
import { LinkInternal } from './LinkInternal.js';
import { LinkMailto } from './LinkMailto.js';
import { MediaProvider, MediaConsumer } from './MediaProvider.js';
import { MenuBar } from './MenuBar.js';
import { MenuDropdown } from './MenuDropdown.js';
import { MenuDropdownIcon } from './MenuDropdownIcon.js';
import { MenuFloating } from './MenuFloating.js';
import { MenuItem } from './MenuItem.js';
import { MenuSeparator } from './MenuSeparator.js';
import { MenuSubmenu } from './MenuSubmenu.js';
import { MessageBox } from './MessageBox.js';
import { Modal } from './Modal.js';
import { PageContent } from './PageContent.js';
import { Pagination } from './Pagination.js';
import { Select } from './Select.js';
import { Tab, TabContainer, TabController, TabPanel } from './TabController.js';
import { Text } from './Text.js';

export { Badge, Box, Button, Divider, FieldCheckbox, FieldElementStyle, FieldError, FieldInfo,
  FieldInput, FieldLabel, FieldSelect, FieldText, FieldTextArea, Flex, FlexItem,
  Font, FontSize, GridBox, GridBoxElement, Image, LinkExternal, LinkInternal, LinkMailto,
  MediaConsumer, MediaProvider, MenuBar, MenuDropdown, MenuDropdownIcon, MenuFloating,
  MenuItem, MenuSeparator, MenuSubmenu, MessageBox, Modal, PageContent, Pagination,
  Select, Tab, TabContainer, TabController, TabPanel, Text };
