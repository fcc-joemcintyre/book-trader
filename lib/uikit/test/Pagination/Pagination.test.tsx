import React from 'react';
import { ThemeProvider } from 'styled-components';
import { cleanup, render } from '@testing-library/react';
import { Pagination } from '../../src/Pagination';
import { theme } from '../util/theme';

afterEach (cleanup);

describe ('Pagination', () => {
  it ('2 elements, first is current', () => {
    const { baseElement } = render (
      <ThemeProvider theme={theme}>
        <Pagination items={20} pageItems={10} visible={5} first={0} current={0} onChange={() => { /* no op */ }} />
      </ThemeProvider>
    );
    expect (baseElement).toMatchSnapshot ();
  });
  it ('2 elements, second is current', () => {
    const { baseElement } = render (
      <ThemeProvider theme={theme}>
        <Pagination items={20} pageItems={10} visible={5} first={0} current={1} onChange={() => { /* no op */ }} />
      </ThemeProvider>
    );
    expect (baseElement).toMatchSnapshot ();
  });
  it ('subset of pages showing, starting at beginning', () => {
    const { baseElement } = render (
      <ThemeProvider theme={theme}>
        <Pagination items={100} pageItems={10} visible={5} first={0} current={0} onChange={() => { /* no op */ }} />
      </ThemeProvider>
    );
    expect (baseElement).toMatchSnapshot ();
  });
  it ('subset of pages showing, showing middle pages', () => {
    const { baseElement } = render (
      <ThemeProvider theme={theme}>
        <Pagination items={100} pageItems={10} visible={5} first={2} current={2} onChange={() => { /* no op */ }} />
      </ThemeProvider>
    );
    expect (baseElement).toMatchSnapshot ();
  });
  it ('subset of pages showing, showing lastpages', () => {
    const { baseElement } = render (
      <ThemeProvider theme={theme}>
        <Pagination items={100} pageItems={10} visible={5} first={5} current={5} onChange={() => { /* no op */ }} />
      </ThemeProvider>
    );
    expect (baseElement).toMatchSnapshot ();
  });
});
