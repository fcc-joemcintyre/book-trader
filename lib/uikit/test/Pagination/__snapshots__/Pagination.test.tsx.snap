// Vitest Snapshot v1

exports[`Pagination > 2 elements, first is current 1`] = `
<body>
  <div>
    <div
      class="sc-bczRLJ sc-gsnTZi bTcrmt dlUnJU"
    >
      <div
        class="sc-dkzDqf hjJIeP"
        cursor="default"
      >
        &lt;
      </div>
      <div
        class="sc-dkzDqf fpkzJG"
        cursor="pointer"
      >
        <strong>
          1
        </strong>
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        2
      </div>
    </div>
  </div>
</body>
`;

exports[`Pagination > 2 elements, second is current 1`] = `
<body>
  <div>
    <div
      class="sc-bczRLJ sc-gsnTZi bTcrmt dlUnJU"
    >
      <div
        class="sc-dkzDqf hjJIeP"
        cursor="default"
      >
        &lt;
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        1
      </div>
      <div
        class="sc-dkzDqf fpkzJG"
        cursor="pointer"
      >
        <strong>
          2
        </strong>
      </div>
    </div>
  </div>
</body>
`;

exports[`Pagination > subset of pages showing, showing lastpages 1`] = `
<body>
  <div>
    <div
      class="sc-bczRLJ sc-gsnTZi bTcrmt dlUnJU"
    >
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        &lt;
      </div>
      <div
        class="sc-dkzDqf fpkzJG"
        cursor="pointer"
      >
        <strong>
          6
        </strong>
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        7
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        8
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        9
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        10
      </div>
    </div>
  </div>
</body>
`;

exports[`Pagination > subset of pages showing, showing middle pages 1`] = `
<body>
  <div>
    <div
      class="sc-bczRLJ sc-gsnTZi bTcrmt dlUnJU"
    >
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        &lt;
      </div>
      <div
        class="sc-dkzDqf fpkzJG"
        cursor="pointer"
      >
        <strong>
          3
        </strong>
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        4
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        5
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        6
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        7
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        &gt;
      </div>
    </div>
  </div>
</body>
`;

exports[`Pagination > subset of pages showing, starting at beginning 1`] = `
<body>
  <div>
    <div
      class="sc-bczRLJ sc-gsnTZi bTcrmt dlUnJU"
    >
      <div
        class="sc-dkzDqf hjJIeP"
        cursor="default"
      >
        &lt;
      </div>
      <div
        class="sc-dkzDqf fpkzJG"
        cursor="pointer"
      >
        <strong>
          1
        </strong>
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        2
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        3
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        4
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        5
      </div>
      <div
        class="sc-dkzDqf iGumvi"
        cursor="pointer"
      >
        &gt;
      </div>
    </div>
  </div>
</body>
`;
